tools/modlog/converter.ts

```diff
@@ -419,12 +421,220 @@ export function rawifyLog(log: ModlogEntry) {
+					userid: result.userid,
+					autoconfirmedID: result.autoconfirmed_userid,
+					alts: result.alts?.split(','),
+					ip: result.ip,
+					isGlobal: result.roomid?.startsWith('global-') || result.roomid === 'global',
+					loggedBy: result.action_taker_userid,
+					note: result.note,
+					time: result.timestamp,
+				};
+
+				const rawLog = rawifyLog(entry);
+				entries.push(rawLog);
+				if (entry.isGlobal) {
+					globalEntries.push(rawLog);
+				}
+				if (entries.length === ENTRIES_TO_BUFFER) await insertEntries();
+			}
+			await insertEntries();
+			if (entriesLogged) process.stdout.write('\n');
+		}
+		if (!Config.nofswriting) console.log(`Writing the global modlog...`);
+		await this.writeFile(`${this.textLogDir}/modlog_global.txt`, globalEntries.join(''));
+	}
+
+	async writeFile(path: string, text: string) {
+		if (this.isTesting) {
+			const old = this.isTesting.files.get(path);
+			return this.isTesting.files.set(path, `${old || ''}${text}`);
+		}
+		return FS(path).append(text);
+	}
+}
+
+export class ModlogConverterTxt {
+	readonly databaseFile: string;
+	readonly database: Database.Database;
+
+	readonly insertionQuery: Database.Statement;
+	readonly FTSInsertionQuery: Database.Statement<[number, string?, string?, string?, string?]>;
+	readonly altsInsertionQuery: Database.Statement<[number, string]>;
+	readonly FTSAltsInsertionQuery: Database.Statement<[number, string]>;
+
+	readonly altsInsertionTransaction: Database.Transaction;
+	readonly insertionTransaction: Database.Transaction;
+	readonly textLogDir: string;
+	readonly isTesting: {files: Map<string, string>, ml?: Modlog} | null = null;
+	constructor(databaseFile: string, textLogDir: string, isTesting?: Map<string, string>, useFTSExtension?: boolean) {
+		this.databaseFile = databaseFile;
+		this.textLogDir = textLogDir;
+		if (isTesting || Config.nofswriting) {
+			this.isTesting = {
+				files: isTesting || new Map<string, string>(),
+			};
+		}
+
+		this.database = Database(this.isTesting ? ':memory:' : this.databaseFile);
+		this.database.exec(FS(`databases/schemas/modlog.sql`).readIfExistsSync());
+		if (useFTSExtension || Config.modlogftsextension) {
+			this.database.exec(`SELECT load_extension('native/fts_id_tokenizer.o')`);
+			this.database.exec(`CREATE VIRTUAL TABLE IF NOT EXISTS modlog_fts USING fts5(note, userid, autoconfirmed_userid, action_taker_userid, content=modlog, content_rowid=modlog_id, tokenize='id_tokenizer')`);
+			this.database.exec(`CREATE VIRTUAL TABLE IF NOT EXISTS alts_fts USING fts5(userid, content=alts, content_rowid=rowid, tokenize='id_tokenizer')`);
+		} else {
+			this.database.exec(`CREATE VIRTUAL TABLE IF NOT EXISTS modlog_fts USING fts5(note, userid, autoconfirmed_userid, action_taker_userid, content=modlog, content_rowid=modlog_id)`);
+			this.database.exec(`CREATE VIRTUAL TABLE IF NOT EXISTS alts_fts USING fts5(userid, content=alts, content_rowid=rowid)`);
+		}
+
+		this.insertionQuery = this.database.prepare(
```

Comment:
The fact that all of the queries are duplicated in the converter suggests that the modlog API is insufficient. The converter is a temporary thing, so I guess this isn't urgent, but it's a red flag

Reply:
oh, I just need to grab the transactions from `Modlog`. That'll be DRYer